Fixed SQL metadata retreival error

object opened the pull request:
The error you experienced yesterady was caused by attempt to treat as numeric value a nullable field in SQL metadata. I have now created an HR schema that you are using to test the provider and run tests for both Oracle client and Devart.

What was changed:
1. Bug fix caused by my changes in DbConverter.
2. HR schema script is revised, foreign key constraints are created after table creation, otherwise script execution fails.
3. LONG datatype mapping is set back to String. Surprisingly enough, Oracle defines LONG as a large text, not as numeric type.
4. Restored proper mapping of NUMBER fields depending on precision and scale.

NB! After I pulled your latest version, I saw that many tests fail. Don't know if they all work in your environment, I am using Oracle 11.2 (not express, standard edition). When applying my changes I verified that it's the same tests that fail before and after the change. That's the best I could do. If the tests work fine on your machine, I suppose they should also work when you apply these changes. If so, let me know, we should find out what makes our environments different, now I have the same HR schema, so everything should be the same.

object committed:
Implemented mapping of Oracle NUMBER and RAW types.

object committed:
Fixed problem with NULL column exception while reading metadata.

flq commented:
Thanks for your work. My suspicion is that Devart's type mapping is a lot better than the Oracle Provider's one. I think I'll have to get the trial version of the devart and see what's happening in direct comparison. It may be that I'll have to extend the compiler flag to the class containing the mapping stuff.

Another question: Did you get my patch? You had removed the data_length which is however relevant for the non-numeric columns and is I think the number that Simple.Data expects when instantiating the column. Also I get one Number column in the HR schema that has no precision info. For this I had made the precision and scale values nullable such that the whole thing can fall through to decimal...

object commented:
The mapping that I added are not Devart specific, although it was Devart developers who studied and published mapping rules. Devart has a trial version of donConnect. If you want to better investigate it I may try to ask them to get you a license for this purpose. It can be in their interest to see Simple.Data support for Devart providers.

What patch do you mean? I just pulled your latest "upgrade" branch from your repository, fixed the bug and re-applied the changes. The reason I removed data_length because I didn't find any cases where it would be used, but if you know such, of course it should be restored.

And what about failing tests? As I wrote earlier, I have failing tests for HR schema both Devart and Oracle. Did you manage to reproduce them?

flq commented:
Simon now keeps the code feeding the official Nuget package at https://github.com/SimonH/Simple.Data.Oracle
